lib/unison-sqlite/src/Unison/Sqlite/DB.hs

```diff
@@ -0,0 +1,272 @@
+-- | A type class interface to SQLite.
+module Unison.Sqlite.DB
+  ( -- * Type-class
+    DB,
+    runDB,
+    runTransaction,
+
+    -- * Executing queries
+
+    -- ** Without results
+
+    -- *** With parameters
+    execute,
+    executeMany,
+
+    -- *** Without parameters
+    execute_,
+
+    -- ** With results
+
+    -- *** With parameters
+    queryList,
+    queryListOne,
+    queryMaybe,
+    queryMaybeOne,
+    queryOne,
+    queryOneOne,
+
+    -- **** With checks
+    queryListCheck,
+    queryListOneCheck,
+    queryMaybeCheck,
+    queryMaybeOneCheck,
+    queryOneCheck,
+    queryOneOneCheck,
+
+    -- *** Without parameters
+    queryList_,
+    queryListOne_,
+    queryMaybe_,
+    queryMaybeOne_,
+    queryOne_,
+    queryOneOne_,
+
+    -- **** With checks
+    queryListCheck_,
+    queryListOneCheck_,
+    queryMaybeCheck_,
+    queryMaybeOneCheck_,
+    queryOneCheck_,
+    queryOneOneCheck_,
+
+    -- * Low-level operations
+    withSavepoint,
+    withStatement,
+  )
+where
+
+import Control.Monad.IO.Unlift (MonadUnliftIO, withRunInIO)
+import Control.Monad.Reader (MonadReader, ReaderT, ask, runReaderT)
+import qualified Database.SQLite.Simple as Sqlite
+import qualified Database.SQLite.Simple.FromField as Sqlite
+import Unison.Prelude
+import Unison.Sqlite.Connection (Connection)
+import qualified Unison.Sqlite.Connection as Connection
+import Unison.Sqlite.Sql (Sql (..))
+import Unison.Sqlite.Transaction (Transaction)
+import qualified Unison.Sqlite.Transaction as Transaction
+
+type DB m =
+  (MonadIO m, MonadReader Connection m)
```

Comment:
I understand the need for the `DB` constraint given the existing uses, but how much work would it be to replace `liftIO` with `transactionalIO` or something similar to give a clear signal that all IO should be repeatable or idempotent?

If that's not feasible, what's the migration/cleanup plan for this?

If there's one thing worse than having a broken abstraction it's having a broken abstraction AND a half migrated one 😋

Reply:
I don't quite understand the suggestion, this interface does not involve transactions and hence does not automatically retry queries, so non-idempotent IO actions are okay. Right? :thinking:

Reply:
I guess I'm not clear on what the plan is for the future of the codebase;
Is this a temporary module which will be cleaned up soon, or are we planning on maintaining both the `DB` version and `Transaction` version indefinitely?